server/src/main/java/org/elasticsearch/snapshots/SnapshotShutdownProgressTracker.java

```diff
@@ -0,0 +1,257 @@
+        @Override
+        public void run() {
+            logProgressReport();
+        }
+    }
+
+    /** How frequently shard snapshot progress is logged after receiving local node shutdown metadata. */
+    public static final Setting<TimeValue> SNAPSHOT_PROGRESS_DURING_SHUTDOWN_INTERVAL_TIME_SETTING = Setting.positiveTimeSetting(
+        "snapshots.shutdown.progress.interval",
+        TimeValue.timeValueSeconds(5),
+        Setting.Property.NodeScope
+    );
+
+    private static final Logger logger = LogManager.getLogger(SnapshotShutdownProgressTracker.class);
+
+    private final ClusterService clusterService;
+    private final ThreadPool threadPool;
+
+    private final ProgressLogger progressLogger = new ProgressLogger();
+    private final TimeValue progressLoggerInterval;
+    private Scheduler.Cancellable scheduledProgressLoggerFuture;
+
+    /**
+     * The time at which the cluster state update began that found a shutdown signal for this node. Negative value means unset (node is not
+     * shutting down).
+     */
+    private volatile long shutdownStartMillis = -1;
+
+    /**
+     * The time at which the cluster state finished setting shard snapshot states to PAUSING, which the shard snapshot operations will
+     * discover asynchronously. Negative value means unset (node is not shutting down)
+     */
+    private volatile long shutdownFinishedSignallingPausingMillis = -1;
+
+    /**
+     * Tracks the number of shard snapshots that have started on the data node but not yet finished.
+     */
+    private final AtomicLong numberOfShardSnapshotsInProgressOnDataNode = new AtomicLong();
+
+    /**
+     * The logic to track shard snapshot status update requests to master can result in duplicate requests (see
+     * {@link ResultDeduplicator}), as well as resending requests if the elected master changes.
+     * Tracking specific requests uniquely by snapshot ID + shard ID de-duplicates requests for tracking.
+     */
+    private final Set<String> shardSnapshotRequests = ConcurrentCollections.newConcurrentSet();
+
+    /**
+     * Track how the shard snapshots reach completion during shutdown: did they fail, succeed or pause?
+     */
+    private final AtomicLong doneCount = new AtomicLong();
+    private final AtomicLong failureCount = new AtomicLong();
+    private final AtomicLong abortedCount = new AtomicLong();
+    private final AtomicLong pausedCount = new AtomicLong();
+
+    public SnapshotShutdownProgressTracker(ClusterService clusterService, Settings settings, ThreadPool threadPool) {
+        this.clusterService = clusterService;
+        this.progressLoggerInterval = SNAPSHOT_PROGRESS_DURING_SHUTDOWN_INTERVAL_TIME_SETTING.get(settings);
+        this.threadPool = threadPool;
+    }
+
+    private void scheduleProgressLogger() {
+        scheduledProgressLoggerFuture = threadPool.scheduleWithFixedDelay(
+            progressLogger,
+            progressLoggerInterval,
```

Comment:
Should we special-case `TimeValue#MINUS_ONE` to mean to disable this feature? At the moment I think that would mean to emit progress logs constantly with zero delay.

Reply:
Yeah I think so. `1ms` is clearly going to cause a tight loop but I think that would be surprising behaviour for `-1` (and arguably for `0` too).

The advantage of having a value which completely disables this functionality is that it gives a potential workaround for any bugs that might crop up later, in a way that running it daily would not.